### PR TITLE
12975 export number of buffered messages as metrics #12975

### LINKED ISSUE #12975: Export number of buffered messages as metrics
**Is your feature request related to a problem? Please describe.**
When configuring async message TTL checking, I need to know the health of my system. A key metric here is the number of "buffered" messages. When this number is steadily climbing, I might need to configure the TTL checker differently to ensure that messages expire fast enough and don't accumulate.

**Describe the solution you'd like**
Export more message related metrics. Similar to the banned instance metrics, these should restore on recovery so that the reported counts are accurate even when brokers restart.

**Describe alternatives you've considered**
Not exposing metrics for this directly, instead relying on the exported record stream. This is more complicated and requires a separate application continously running.

**Additional context**

related to https://jira.camunda.com/browse/SUPPORT-17177

### PR DESCRIPTION
**Pull request by rodrigo-lourenco-lopes**
## Description

Add metrics for Column Family Entries.
In Grafana a panel under the Messages, namely Buffered messages is counting the number of current message deadlines.

image

## Related issues

closes #12975

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

**Review by deepthidevaki — CHANGES_REQUESTED** [review]
Thanks @rodrigo-lourenco-lopes 
Besides my comments:

- :x: The issue also says to recover the count after restart. With the current PR the total count will be reset to 0 after a pod restart. :thinking: Recovering the count after restart will be an expensive operation, since you have to iterate over all keys in the database. We should discuss and agree what is the expected behavior. If it is just to keep the count of messages, we can also keep the count in rocksdb itself, and read it after restart. But it might not be valuable to keep the count for all columnfamilies. @Zelldon @oleschoenburg  Any thoughts on this?

**Comment by Zelldon** [comment]
I don't know what was discussed with @oleschoenburg, but I also thought we store the counts.

What I was imagining was having an extra column family (STATS) or using the default one or having the counter always in the corresponding CF and storing all counts of each column family. These can then be incremented when new data is inserted or decremented if something is deleted. No iteration etc needed. We can then expose all of these counters via metrics (gauge). 

Of course we can argue whether we really want to have that for everything, but might simplify the implementation IDK haven't looked into it in detail.

**Comment by deepthidevaki** [comment]
> What I was imagining was having an extra column family (STATS) or using the default one 

This would be a good approach.

However, I'm worried about the performance impact of queries like 
```
 if (!exists(key)) {
   // increment count
} // then put

 if (exists(key)) {
   //decrement  count
} // then delete
```
This check is necessary to ensure the count is consistent with the state.

**Comment by Zelldon** [comment]
@deepthidevaki but we could read the count once right on restart and have this in memory and reflect all updates with a simple update.

**Comment by deepthidevaki** [comment]
I'm not talking about recovering the count after restart. But, how we update the count for every upsert and delete. See the changes in TransactionalColumnFamily in this PR.

**Comment by Zelldon** [comment]
> I'm not talking about recovering the count after restart. But, how we update the count for every upsert and delete. See the changes in TransactionalColumnFamily in this PR.

Yeah I know but still my comment applies. I don't think we need to have the checks then.

**Comment by deepthidevaki** [comment]
> > I'm not talking about recovering the count after restart. But, how we update the count for every upsert and delete. See the changes in TransactionalColumnFamily in this PR.
> 
> Yeah I know but still my comment applies. I don't think we need to have the checks then.

Why not? What if we are updating an already existing key? Then without the check we will count it twice. Same for delete, if we delete the same key twice this can result in wrong count.

**Comment by rodrigo-lourenco-lopes** [comment]
Regarding the performance of the current implementation, I created benchmarks with mixed workload and also another one for reference from main, and I was not able to find any difference in performance. But perhaps we can have a feature flag or configuration for this column family metrics?

**Comment by Zelldon** [comment]
> Why not? What if we are updating an already existing key? Then without the check we will count it twice. Same for delete, if we delete the same key twice this can result in wrong count.

Ok I think we were talking about two different things.

I was talking about checking whether the count exists and not rereading the count value. And you were talking about whether the actual key of the column family already exists, e.g like a process instance.


I think we can completely integrate the counts in the column family implementation, where we already have the checks for existing. But yes I agree that if we rely on these checks then this is a performance issue, especially on large state as we have seen in my benchmarks with large state.

@rodrigo-lourenco-lopes yes feature flag would be 100% something we should do.

**Comment by deepthidevaki** [comment]
@rodrigo-lourenco-lopes You might see a performance impact if the state is large. To check if a key exists, it sometimes have to read multiple disk i/o especially when the key does not exist.

> I think we can completely integrate the counts in the column family implementation, where we already have the checks for existing

@Zelldon I don't understand your comment. We have to introduce checks at two new places (in upsert and delete) for these metrics.

**Comment by Zelldon** [comment]
I see. Yeah I guess might be not optimal, to always check the existence, but I have no numbers for it. Could also be cached already so might be fine (bloom-filter, cache lot of things are in rocksdb might cover this). We could do some more extensive testing to find this out. With large state etc. You could also rebase to main then you have the engine performance test available.

In general I guess would be great to have the insert method back again, which is use in code places where we know that the instance or the message doesn't exist and we are creating it. In this case we can count such easily.

Tbh I'm not sure that we need to make it too complicated the metrics should be just a rough estimate, we don't need to be precise, I guess. We just want to understand do we have around 1000 instances or buffered messages or do we have millions.

Not sure whether we need to check at all in this case. We could also just count upserts and deletes and compare the diff.  Wouldn't that be enough?

If we want to count more exactly we could reset it always on restart via iterating over the CF :shrug:

**Comment by Zelldon** [comment]
We can also discuss this f2f

**Comment by deepthidevaki** [comment]
> We can also discuss this f2f

:+1: @rodrigo-lourenco-lopes Could you schedule a slot? Let's unblock this PR asap.

File: engine/src/main/java/io/camunda/zeebe/engine/state/message/DbMessageState.java [hunk 1]
```diff
@@ -18,14 +18,18 @@
...
 import io.camunda.zeebe.engine.state.mutable.MutableMessageState;
 import io.camunda.zeebe.protocol.ZbColumnFamilies;
 import io.camunda.zeebe.protocol.impl.record.value.message.MessageRecord;
+import io.camunda.zeebe.stream.api.ReadonlyStreamProcessorContext;
 import org.agrona.DirectBuffer;
 import org.agrona.collections.MutableBoolean;
 
 public final class DbMessageState implements MutableMessageState {
 
+  private static final Long DEADLINE_MESSAGE_COUNT_KEY = 1L;
```
Comment (deepthidevaki) [inline engine/src/main/java/io/camunda/zeebe/engine/state/message/DbMessageState.java:31]: :wrench: IMO this is a bit weird to be 1 as the key.  I would recommend to have a more general stats column family and specific keys.

```
private static final String DEADLINE_MESSAGE_COUNT_KEY = "deadline_message_count";
private final ColumnFamily<DbString, DbLong> messageStatsColumnFamily;

....
  messagesDeadlineCount = new DbLong();
  messagesDeadlineCountKey = new DbString();
  messageStatsColumnFamily =
      zeebeDb.createColumnFamily(
  ZbColumnFamilies.MESSAGE_STATS,
  transactionContext,
  messagesDeadlineCountKey,
  messagesDeadlineCount);

  messagesDeadlineCountKey.wrapString(DEADLINE_MESSAGE_COUNT_KEY);
```

Or have a generic column family for such metrics, so that we can add more metrics in future.

```
  statsColumnFamiliy =
      zeebeDb.createColumnFamily(
  ZbColumnFamilies.STATS_COUNT,
  transactionContext,
  messagesDeadlineCountKey,
  messagesDeadlineCount);
```

What do you think?

File: zb-db/src/main/java/io/camunda/zeebe/db/impl/rocksdb/BufferedMessagesMetrics.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+import io.prometheus.client.Gauge;
+
+public class BufferedMessagesMetrics {
+
+  private static final Gauge BUFFERED_MESSAGES_COUNT =
+      Gauge.build()
+          .namespace("zeebe")
+          .name("buffered_messages_count")
+          .help("Current number of buffered messages entries.")
```
Comment (deepthidevaki) [inline zb-db/src/main/java/io/camunda/zeebe/db/impl/rocksdb/BufferedMessagesMetrics.java:18]:
```suggestion
          .help("Current number of buffered messages.")
```

File: protocol/src/main/java/io/camunda/zeebe/protocol/ZbColumnFamilies.java [hunk 3]
```diff
@@ -15,6 +15,8 @@
  */
 package io.camunda.zeebe.protocol;
 
+// New Column families should be added at the bottom as ColumnFamilyContext uses
+// this class as ordinal()
```
Comment (deepthidevaki) [inline protocol/src/main/java/io/camunda/zeebe/protocol/ZbColumnFamilies.java:19]: :+1:

File: engine/src/test/java/io/camunda/zeebe/engine/state/ProcessExecutionCleanStateTest.java [hunk 4]
```diff
@@ -220,7 +221,7 @@ public void testProcessWithMessageEvent() {
             .create();
 
     // when
-    final var timeToLive = Duration.ofSeconds(10);
+    final var timeToLive = Duration.ofSeconds(100);
```
Comment (deepthidevaki) [inline engine/src/test/java/io/camunda/zeebe/engine/state/ProcessExecutionCleanStateTest.java:224]: :question: Why is this change necessary?

File: engine/src/main/java/io/camunda/zeebe/engine/state/message/DbMessageState.java [hunk 5]
```diff
@@ -180,6 +221,18 @@ public void put(final long key, final MessageRecord record) {
     deadline.wrapLong(record.getDeadline());
     deadlineColumnFamily.insert(deadlineMessageKey, DbNil.INSTANCE);
 
+    long count = 0L;
+    if (!messagesDeadlineCountColumnFamily.isEmpty()) {
+      count = messagesDeadlineCountColumnFamily.get(messagesDeadlineCountKey).getValue();
+      messagesDeadlineCount.wrapLong(count + 1L);
+      messagesDeadlineCountColumnFamily.update(messagesDeadlineCountKey, messagesDeadlineCount);
+    } else {
```
Comment (deepthidevaki) [inline engine/src/main/java/io/camunda/zeebe/engine/state/message/DbMessageState.java:229]: :x: We can keep an in-memory `count` and keep it updated, instead of `get` every time. Then no need to check if the key already exists. `get` can be expensive it has to read data that is not in memory.

File: engine/src/main/java/io/camunda/zeebe/engine/state/message/DbMessageState.java [hunk 6]
```diff
@@ -180,6 +221,18 @@ public void put(final long key, final MessageRecord record) {
...
+      messagesDeadlineCount.wrapLong(count + 1L);
+      messagesDeadlineCountColumnFamily.insert(messagesDeadlineCountKey, messagesDeadlineCount);
```
Comment (deepthidevaki) [inline engine/src/main/java/io/camunda/zeebe/engine/state/message/DbMessageState.java:231]: :x:
```suggestion
      messagesDeadlineCountColumnFamily.upsert(messagesDeadlineCountKey, messagesDeadlineCount);
```
insert will fail if the key already exists.

File: monitor/grafana/zeebe.json [hunk 7]
```diff
@@ -1,83 +1,4 @@
...
-      "name": "Table (old)",
-      "version": ""
-    },
-    {
-      "type": "panel",
-      "id": "timeseries",
-      "name": "Time series",
-      "version": ""
-    }
-  ],
```
Comment (deepthidevaki) [inline monitor/grafana/zeebe.json:80]: I think these changes have to be reverted.

File: monitor/grafana/zeebe.json [hunk 8]
```diff
@@ -6588,6 +6520,100 @@
           ],
           "title": "Request count per type",
           "type": "timeseries"
+        },
+        {
+          "datasource": {
+            "type": "prometheus",
+            "uid": "${DS_PROMETHEUS}"
+          },
+          "description": "Shows the number of buffered messages (deadline messages) per partition.",
```
Comment (deepthidevaki) [inline monitor/grafana/zeebe.json:6529]:
```suggestion
          "description": "Shows the number of buffered messages per partition.",
```

**Review by deepthidevaki — CHANGES_REQUESTED** [review]
:rocket: Greate.

I have a few more small comments. I will approve the PR after we discuss them. Please also re-do grafana dashboard as there is a conflict with `main`.

File: engine/src/main/java/io/camunda/zeebe/engine/state/message/DbMessageState.java (hunk 1 above)
Comment (rodrigo-lourenco-lopes) [reply engine/src/main/java/io/camunda/zeebe/engine/state/message/DbMessageState.java:31]: I think having a general MESSAGE_STATS is good as it can be any type of stat and not just count 🤔

Comment (deepthidevaki) [reply engine/src/main/java/io/camunda/zeebe/engine/state/message/DbMessageState.java]: >  it can be any type of stat and not just count

Remember that the type of the value in a columnfamily is pre-defined. So you cannot store anything in a general MESSAGE_STATS, but only numbers.

**Review by deepthidevaki — CHANGES_REQUESTED** [review]
:rocket: Thanks for addressing my comments. The code changes looks good. 

I have only some minor comments.
1. :x: Changes to grafana dashboard reverted some of the changes from the main. (Past week I have added a new section for journal, it is now removed in this PR). Please take the latest version from main as the base and add new panels there.
2. :x: Move the panel for buffered messages to section "Processing", may be next to banned instances. Currently it is in "Messaging" which actually is related to network messages. Buffered message are not related to network messages.
3. :wrench: Please cleanup the commit history. Try to rebase on main instead for merging main to your branch. This will help to have a clean commit history in your PR. You can also remove the issue id from the PR title.

**Review by deepthidevaki — APPROVED** [review]
:tada:

**Comment by rodrigo-lourenco-lopes** [comment]
bors merge
